core/src/poh_recorder.rs

```diff
@@ -537,4 +615,104 @@ mod tests {
         // Make sure the starting slot is updated
         assert_eq!(poh_recorder.start_slot(), end_slot);
     }
+
+    #[test]
+    fn test_reached_leader_tick() {
+        let (genesis_block, _mint_keypair) = GenesisBlock::new(2);
+        let bank = Arc::new(Bank::new(&genesis_block));
+        let prev_hash = bank.last_blockhash();
+        let (mut poh_recorder, _entry_receiver) =
+            PohRecorder::new(0, prev_hash, 0, None, bank.ticks_per_slot());
+
+        // Test that with no leader slot, we don't reach the leader tick
+        assert_eq!(poh_recorder.reached_leader_tick(), false);
+
+        for _ in 0..bank.ticks_per_slot() {
+            poh_recorder.tick();
+        }
+
+        // Tick should not be recorded
+        assert_eq!(poh_recorder.tick_height(), 0);
+
+        // Test that with no leader slot, we don't reach the leader tick after sending some ticks
+        assert_eq!(poh_recorder.reached_leader_tick(), false);
+
+        poh_recorder.reset(
+            poh_recorder.tick_height(),
+            bank.last_blockhash(),
+            0,
+            None,
+            bank.ticks_per_slot(),
+        );
+
+        // Test that with no leader slot in reset(), we don't reach the leader tick
+        assert_eq!(poh_recorder.reached_leader_tick(), false);
```

Comment:
do you have this test twice?

Reply:
The part of test is repeated. In first case it's testing that doing reset removes wait for grace ticks. The second case waits for the grace ticks to expire.